File: Advanced Techniques/Substring_Reversals.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace __gnu_pbds;
using namespace std;

#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
#define len(arr) arr.size()
#define f first
#define s second
#define pb push_back
#define all(x) x.begin(), x.end()
#define range(i, n) for (int i = 0; i < (n); ++i)
#define rangea(i, a, b) for (int i = (a); i < (b); ++i)
#define int long long
#define MOD1 998244353
#define MOD 1000000007
using vi = vector<int>;
using vii = vector<vi>;
using viii = vector<vii>;
using ll = long long;
using pi = pair<int,int>;
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")
mt19937 rng(chrono::steady_clock::now().time_since_epoch().count());

#ifdef LOCAL
#define dbg(x...) cout<<"["<<#x<<"] = "; print(x);
#else
#define dbg(x...)
#endif

template <typename T>
using ordered_multiset = tree<
    std::pair<T, int>,
    null_type,
    std::less<std::pair<T, int>>,
    rb_tree_tag,
    tree_order_statistics_node_update
>;
template<class T> bool chmin(T& a,const T& b){ if(b<a){ a=b; return 1;} return 0; }
template<class T> bool chmax(T& a,const T& b){ if(a<b){ a=b; return 1;} return 0;}
template<class T> using minpq = priority_queue<T,vector<T>,greater<T>>;
template<class T> using maxpq = priority_queue<T>;
template<typename T> istream& operator>>(istream& in, vector<T>& v) {for (auto& x : v) in >> x;return in;}
template<typename T1, typename T2> istream& operator>>(istream& in, pair<T1, T2>& p) {in >> p.first >> p.second;return in;}
template<class T> void _print(const T &x){ cout << x; }
template<class T,class U> void _print(const pair<T,U> &p){cout<<"(";_print(p.first);cout<<",";_print(p.second);cout<<")";}
template<class T>
void _print(const vector<T> &v){
    cout<<"[";bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"]";
}
template<class T>
void _print(const set<T> &v){
    cout<<"{";
    bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"}";
}
template<class T>
void _print(const unordered_set<T> &v){
    cout<<"{";
    bool f=0;
    for(const auto &x: v){
        if(f) cout<<", ";
        _print(x);
        f=1;
    }
    cout<<"}";
}
template<class T,class U>
void _print(const unordered_map<T,U> &m){
    cout<<"{";
    bool f=0;
    for(const auto &p: m){
        if(f) cout<<", ";
        _print(p);
        f=1;
    }
    cout<<"}";
}
template<class T,class U>
void _print(const map<T,U> &m){
    cout<<"{";
    bool f=0;
    for(const auto &p: m){
        if(f) cout<<", ";
        _print(p);
        f=1;
    }
    cout<<"}";
}
void print(){ cout<<"\n"; }
template<class T,class... Args>
void print(const T& a,const Args&... args){
    _print(a);
    if constexpr(sizeof...(args)) cout<<" ";
    print(args...);
}

int gcd(int a, int b) {if (b == 0) return a;return gcd(b, a % b);}
int lcm(int a, int b) {return a / gcd(a, b) * b;}
int II(){int a;cin>>a;return a;}
string SI(){string s;cin>>s;return s;}
vi LII(int n){vi a(n);cin>>a;return a;}
int rnd(int l,int r){return uniform_int_distribution<int>(l,r)(rng);}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable, SortedList: sortl
// Segment Tree(lazy propogation): SegmentTree, Merge Sort Tree: sorttree, Trie/Treap: Tries
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, Centroid Decomp.: graph_decom
// Persistent Segment Tree: perseg, FreqGraphs: bgraph, GrapthOth: graphoth, DSU: DSU, FFT:fft
// Rollback/Par DSU: rbdsu, treap: treap, graphflow(mat_match): graphflow, Persistent Seg Tree: perseg
// Segment Tree(Nodes): SegmentNode, HLD: hld, fwht: fwht
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

struct Node {
    int val, pr, sz,rev;
    Node *l, *r;
    Node(int v) : val(v), pr(rng()), sz(1), l(nullptr), r(nullptr),rev(0) {}
};

int sz(Node* t) { return t ? t->sz : 0; }

void upd(Node* t) {
    if (t) t->sz = 1 + sz(t->l) + sz(t->r);
}

Node* merge(Node* a, Node* b) {
    if (!a || !b) return a ? a : b;

    if (a->pr > b->pr) {
        if(a->rev){
            swap(a->r,a->l);
            a->rev = 0;
            if(a->r)a->r->rev ^= 1;
            if(a->l)a->l->rev ^= 1;
        }
        a->r = merge(a->r, b);
        upd(a);
        return a;
    } else {
        if(b->rev){
            swap(b->r,b->l);
            b->rev = 0;
            if(b->r)b->r->rev ^= 1;
            if(b->l)b->l->rev ^= 1;
        }
        b->l = merge(a, b->l);
        upd(b);
        return b;
    }
}

void rev(Node* a){
    a->rev ^= 1;
}

pair<Node*,Node*> split(Node* t, int k) {
    if (!t) return {nullptr,nullptr};
    if(t->rev){
        swap(t->r,t->l);
        t->rev = 0;
        if(t->r)t->r->rev ^= 1;
        if(t->l)t->l->rev ^= 1;
    }
    if (sz(t->l) >= k) {
        auto p = split(t->l, k);
        t->l = p.second;
        upd(t);
        return {p.first, t};
    } else {
        auto p = split(t->r, k - sz(t->l) - 1);
        t->r = p.first;
        upd(t);
        return {t, p.second};
    }
}

void inorder(Node* t, string &out) {
    if (!t) return;
    if(t->rev){
        swap(t->l,t->r);
        if(t->l)t->l->rev ^= 1;
        if(t->r)t->r->rev ^= 1;
    }
    inorder(t->l, out);
    out.pb(t->val);
    inorder(t->r,out);
}

void solve() {
    int n,m;
    cin>>n>>m;
    string st;
    cin>>st;

    Node* st1 = new Node(st[0]);
    for(int i=1;i<n;i++){
        Node* st2 = new Node(st[i]);
        st1 = merge(st1,st2);
    }

    auto inord = [](Node* st){
        string out;
        inorder(st,out);
        print(out);
    };


    while(m--){
        int l,r;
        cin>>l>>r;
        l--;
        r--;
        auto [node1,node2] = split(st1,l);
        auto [node3,node4] = split(node2,(r-l)+1);
        rev(node3);
        st1 = merge(merge(node1,node3),node4);
    }
    // print(1);
    inord(st1);
}

int32_t main() {
    fastio
    int t=1;
    // cin >> t;
    while (t--) solve();
    return 0;
}

File: Advanced Techniques/Substring_Reversals.py
import sys,math,cmath,random,os
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations
from io import BytesIO, IOBase
 
BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
 
# functions #
# MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(1,2**62)
def gcd(a,b):
    if a%b==0:
        return b
    else:
        return gcd(b,a%b)
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
##
 
#String hashing: sh/shclass, fenwick sortedlist: fsortl, Number: numtheory, SparseTable: SparseTable
#Bucket Sorted list: bsortl, Segment Tree(lazy propogation): SegmentTree/Other, bootstrap: bootstrap
#binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, Trie: Tries
#Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, DSU: DSU, Geometry: Geometry, FFT: fft
#Persistent Segment Tree: perseg, FreqGraphs: bgraph, Binary Trie: b_trie, XOR_dict: xdict, HLD: hld
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
# input_file = open(r'input.txt', 'r');sys.stdin = input_file
 
class Treap:
    class Node:
        __slots__ = ('val','prio','left','right','size','rev')
        def __init__(self,val):
            self.val=val
            self.prio=random.random()
            self.left=None
            self.right=None
            self.size=1
            self.rev=False
 
    def __init__(self,s=""):
        """Build treap from string s in O(n) by repeated merges"""
        self.root=None
        for ch in s:
            node=self.Node(ch)
            self.root=self._merge(self.root,node)
 
    def _update(self,node):
        node.size=1
        if node.left: node.size+=node.left.size
        if node.right: node.size+=node.right.size
 
    def _push(self,node):
        if node and node.rev:
            node.rev=False
            node.left, node.right = node.right, node.left
            if node.left: node.left.rev=not node.left.rev
            if node.right: node.right.rev=not node.right.rev
 
    def _split(self,node,count):
        """Iterative split: first count into left"""
        if not node: return (None,None)
        self._push(node)
        left_size = node.left.size if node.left else 0
        if count<=left_size:
            # split left subtree
            l,l2=self._split(node.left,count)
            node.left=l2
            self._update(node)
            return (l,node)
        else:
            # split right subtree
            r1,r=self._split(node.right,count-left_size-1)
            node.right=r1
            self._update(node)
            return (node,r)
 
    def _merge(self,a,b):
        """Iterative merge using loop"""
        if not a or not b: return a or b
        # choose root
        if a.prio>b.prio:
            root=a
            self._push(root)
            root.right=self._merge(root.right,b)
        else:
            root=b
            self._push(root)
            root.left=self._merge(a,root.left)
        self._update(root)
        return root
 
    def reverse(self,l,r):
        """Reverse substring [l,r) 0-indexed"""
        A,B=self._split(self.root,l)
        M,C=self._split(B,r-l)
        if M: M.rev=not M.rev
        self.root=self._merge(self._merge(A,M),C)
 
    def __repr__(self):
        """Iterative inorder traversal"""
        res=[]
        stack=[]
        node=self.root
        while stack or node:
            while node:
                self._push(node)
                stack.append(node)
                node=node.left
            node=stack.pop()
            res.append(node.val)
            node=node.right
        return ''.join(res)
def solve():
    n,m = list(map(int, sys.stdin.readline().split()))
    #L1 = list(map(int, sys.stdin.readline().split()))
    st = sys.stdin.readline().strip()
    tp = Treap(st)
    for i in range(m):
        l,r = list(map(lambda x:int(x)-1, sys.stdin.readline().split()))
        tp.reverse(l,r+1)
    print(tp)        
solve()
